New feature: Upload measured concentrations for samples

leandroradusky opened the pull request:
Closes #1618.

Users can now upload results for samples. While the behavior and designs are implemented, there are several things to comment, review, and define here or in new issues. 

**Screens**

- This action is done from the samples index:

![image](https://user-images.githubusercontent.com/13782680/203114680-94e922aa-449a-4555-90ea-a1b759a107bd.png)
(Mockup: https://projects.invisionapp.com/d/main#/console/4103589/453801736/preview)

- The user access to the upload results page which looks like this:

![image](https://user-images.githubusercontent.com/13782680/203321828-359148c3-c263-416a-84b2-a2d567298c72.png)

- If there are unrecognized samples, a popup will show them:

![image](https://user-images.githubusercontent.com/13782680/203123620-c1b9527f-2628-488d-9845-d439b7846978.png)
(Mockup: https://projects.invisionapp.com/d/main#/console/4103589/470792645/preview)

Here, I've defined the tooltip to show `onmouseover` and disappear `onmouseleave`. Seeing the mockup, I'm not sure if it should behave like this, or `onclick` with some time to disappear. 

**Respect to the code**

- To validate the existence of the samples in the uploaded files, I've created a `existing_uuids` method to be called from the client side. That will retrieve the found uuids. The code for this is in `_upload_results_js.haml`. While the code work ok the final result is quite sophisticated don't know if can be done in a clearer way.
- The `checkmark-icon` for boxes with all its uuids recognized was defined by me since I didn't found it defined, I'm not sure if its just that I didn't found it.
- The way to access `samples/upload_results` in `routes.rb` works ok but I'm not sure this is the right way to define it.
- I've made specs to verify that measurements are defined and not overwritten when re-uploaded.

**Respect to the accepted values in the csv**

We're parsing the uploaded CSVs to accept: `float` numbers (raw or scientific notation) for `measured_signal` and `"positive"` or `"negative"` strings for `measurement_result` fields. This processing is done after the user confirms the upload, if a number/value is invalid, the results are just not uploaded, and this is not informed to the user at the moment.

We should:
- validate the file on the client side (javascript)?
- inform the failure after, with a `notice` message? 
- don't inform it, just be extra clear in the confirmation popup before confirming?
- Come back to the same screen, don't upload and inform the problem?

**Respect to the confirmation popup**

The only screen that doesn't look like its [mockup](https://projects.invisionapp.com/d/main#/console/4103589/470792646/preview) is the upload confirmation mockup, which looks like this:

![image](https://user-images.githubusercontent.com/13782680/203070129-da7dc61b-f831-4f7b-b966-bb5a6486b507.png)

Things to mention: 

- The `confirmation_modal` that already exists in the project actually has an attribute called `deletion` that will modify its style and make it look red, but also this will force its title to be `Delete confirmation` (`Confirm` otherwise). Modifying this component to allow the `Warning` title plus custom styling appears to deserve its own issue. If there is a better way to be resolved within this PR that I'm not realizing please let me know. 
- I added to the warning message the sentence: `Samples with uploaded measurements won't be overwritten`. This is in order to consider two use cases: 1) actual, as mentioned above we're checking for unexistent UUIDs to inform the user, but we're not considering valid samples with uploaded messages, we

**Respect to samples unblinding after uploading and showing the uploaded results**

In the issue it says: After uploading, we show the measured concentrations / results as a column. See the [invision](https://projects.invisionapp.com/d/main#/console/4103589/470303882/preview). 

This was valid when results were uploaded for boxes, but now they are uploaded for samples (could belong or not to one or many boxes). We can add conditionally the measurements in the form, and unblind the sample accordingly if the measurements are uploaded (both? just one?). This is defined to be done in #1735, there is some description there but appears to apply to the "upload samples for a box" old behavior and the policy needs to be redefined.

ysbaddaden commented on app/views/samples/_upload_results_js.haml:
```diff
@@ -0,0 +1,137 @@
+                            .append( `${samples.length} Samples, (${(samples.length-found.length)} sample UUID(s) not found) ` ) 
+                            .append( $('<div>', { class: `icon-alert icon-red bigger` }) )
+                        )
+                    );
```
This could be abstracted with a `<template>` element. You'd only have to set a few values in the cloned DOM. See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template

```html
<template id="csvFileRow">
  <div class="">
    <span class="samples-count"></span> Samples
  </div>
</template>
```

```js
var template = document.getElementById("csvFileRow").content;
var fragment = template.cloneNode(true);
fragment.querySelector(".samples-count").textContent = samples.length;
```

ysbaddaden replied:
Note that you don't have to use jQuery. The DOM is nice/standard enough nowadays.

ysbaddaden reviewed (commented):
I left some comments to improve readability.

The one thing that annoys me is as often the sheer amount of JavaScript. Maybe it could be simplified a bit using the `<template>` element?

diegoliberman commented:
> Here, I've defined the tooltip to show onmouseover and disappear onmouseleave. Seeing the mockup, I'm not sure if it should behave like this, or onclick with some time to disappear.

@leandroradusky this UX/UI questions go for the designer as a first option, @jkicillof in this case.  It isn't intuitive that there is tooltip there, IMO.

diegoliberman commented:
> The checkmark-icon for boxes with all its uuids recognized was defined by me since I didn't found it defined, I'm not sure if its just that I didn't found it.

@jkicillof where should we get the icon from? cc @leandroradusky

diegoliberman commented:
>  if a number/value is invalid, the results are just not uploaded, and this is not informed to the user at the moment.

I think we should show the error message in this screen: https://projects.invisionapp.com/d/main#/console/4103589/470842165/preview

1. We need to include the list of IDs with the problem, with a maximum of 5 IDs.  
2. The same goes for the IDs not found in CDx, we should list them, max 5.

diegoliberman commented:
> Modifying this component to allow the Warning title plus custom styling appears to deserve its own issue.

I totally agree.  You can create the issue within the increment, if we have time by the end of the project, we'll try to implement it.

diegoliberman commented:
> I added to the warning message the sentence: Samples with uploaded measurements won't be overwritten. This is in order to consider two use cases: 1) actual, as mentioned above we're checking for unexistent UUIDs to inform the user, but we're not considering valid samples with uploaded messages, we

I agree, but your last sentence seems truncated @leandroradusky

diegoliberman commented:
> This was valid when results were uploaded for boxes, but now they are uploaded for samples (could belong or not to one or many boxes). We can add conditionally the measurements in the form, and unblind the sample accordingly if the measurements are uploaded (both? just one?). This is defined to be done in https://github.com/instedd/cdx/issues/1735, there is some description there but appears to apply to the "upload samples for a box" old behavior and the policy needs to be redefined.

1. A sample Can't belong to many boxes.  Right?
2. If the sample gets uploaded a result, whether the boolean value or the measured signal, then the samples gets unblinded.  Let's keep it simple, we can change it later if we need a more complex logic.  
3. I don't quite follow why the box has anything to do here.  I suspect that this might be coming from the fact that the Blind attribute belongs to the box entity, instead of the sample entity.  If that's the case, I think we might have no choice but to move the attribute to the Sample entity.  But I'd like to know what @ysbaddaden thinks.

cc @leandroradusky

jkicillof commented:
https://github.com/instedd/cdx/pull/1805#issuecomment-1324150379
@leandroradusky The idea was displaying the UUID list on click, the designs show a dashed underline there.

jkicillof commented:
>The checkmark-icon for boxes with all its uuids recognized was defined by me since I didn't found it defined, I'm not sure if its just that I didn't found it.

The icons are mostly from [Google material](https://fonts.google.com/icons).
![check](https://user-images.githubusercontent.com/5031636/203541646-24b1ef89-a398-42c1-a1b1-c73983a7c8cc.svg)

leandroradusky commented:
Fixes suggested for the PR review were made. Some comments below:

- While now `measured_signal` and `measurement_result` fields are correctly defined in the `core_fields` definition, the validations remain: for the `float` field, they don't work (whatever string would be saved); for the enum field the error is unfriendly (failure red page) and I added downcasing to allow the input csv to tolerate upper and lower. Friendlier messages will be displayed when #1807 & #1808 are implemented.

- I removed all the jQuery and implemented the `<template>` solution, I defined the templates in haml. The resulting code isn't shorter (and IMO not clearer, suggestions are always welcome, cannot achieve prettiness on the javascript and I'm not happy with it).  

- Respect to the checkmark icon, I was referring to the `css` class to define it, as is defined for the other icons (`icon-warning`, `icon-plus`, etc). Its being defined like the others and taken from the glyphs svg then is well defined, just was asking if maybe was already defined with other name than checkmark and I was repeating code without realizing but is not the case.

- I have corrected the display message to add the dashed underline and display `onclick`. The tooltip will disappear on re-click or in 5 seconds. 

- Two new issues were created to handle the remaining commentary (show the proper messages in the modal popup): we need a component similar to the `ConfirmationModal` allowing display of custom messages first (#1807). Then we will be able to properly show the messages detailed in the mockups(#1808). The actual component is programmed to confirm and confirm deletion and all its logit its wired in this way to apply styles, titles and messages.

ysbaddaden reviewed (commented):
@leandroradusky here are a few ideas to start simplifying the JS.

leandroradusky commented on app/views/samples/_upload_results_js.haml:
```diff
@@ -57,52 +58,53 @@
-                    })
+                    document.getElementById(`not_found_message_${now}`).addEventListener('click', 
+                        function() {
+                            if (!document.getElementById(`ttext_${now}`)){
+                                var tooltip = document.createElement("span")
+                                tooltip.className = `ttext`
+                                tooltip.id = `ttext_${now}`
+                                tooltip.style = `color:red; width: 300px;`
+                                tooltip.innerHTML = `${not_found.slice(0,5).join("<br>")}`
+                                document.getElementById(`not_found_message_${now}`).append(tooltip)
+                                setTimeout( () => tooltip.remove() , 5000 );
+                            }
+                            else {
+                                document.getElementById(`ttext_${now}`).remove()
+                            }
```
The problem here is that this is within the function added as click event listener and we don't have access to `fragment` there. I'll check if this can be solved, but was the first thing I tried and didn't work.

leandroradusky replied:
Even if I do 

```
fragment.querySelector(".not_found_message").fragment = fragment
fragment.querySelector(".not_found_message").addEventListener('click', clickF, false )
```

and then 

```
var clickF = function(evt) {
       console.log(evt.currentTarget.fragment)
       var tooltip = evt.currentTarget.fragment.querySelector(".ttip")
       console.log(tooltip)
```

I would get a null for the query selector :S 

![image](https://user-images.githubusercontent.com/13782680/203826596-e5690144-2698-47aa-af67-57a55f46bf06.png)

leandroradusky replied:
It works passing the tooltip (the result of the querySelector) as param instead of the fragment :)

leandroradusky commented on app/views/samples/_upload_results_js.haml:
```diff
@@ -115,17 +117,11 @@
-        $('#uploaded-files').append(new_file);
+        var template = document.getElementById(`csvInputFile`).content;
+        var fragment = template.cloneNode(true);
+        fragment.querySelector(`.csv_file`).id = `csv_file_${now}`;
```
This won't work since the DOM isn't ready and `csvFile` will be null, this obliged me to keep the now param in this place only... Any suggestion?

ysbaddaden replied:
AFAIK the DOM objects won't change (you can add/remove them at will), so you can keep the reference?

```js
var fileInput = fragment.querySelector('.csv_file');
document.getElementById('uploaded-files').append(fragment);
fileInput.addEventListener('change', handleFileSelect, false);
fileInput.click();
```

leandroradusky replied:
It worked! :D

ysbaddaden commented on app/views/samples/_upload_results_js.haml:
```diff
@@ -0,0 +1,125 @@
+            return async(load_event) => {
+                // Parse the input file
+                var text =  load_event.target.result
+                var lines = text.split(/\r\n|\r|\n/); // tolerate both Windows and Unix linebreaks
```
:+1: for allowing CRLF, CR and LF

ysbaddaden commented on app/views/samples/_upload_results_js.haml:
```diff
@@ -0,0 +1,125 @@
+                fragment.querySelector(`.file-name`).textContent = `${f.name}`;
+                // if samples were not found add mouse behaviour to display tooltip
+                if (found.length != samples.length) {
+                    var showToolTip = function(evt) {
+                            var tooltip = evt.currentTarget.tooltip
+                            if (tooltip.classList.contains("hidden")) {
+                                tooltip.classList.remove("hidden")
+                                setTimeout(() => tooltip.classList.add("hidden"), 5000)
+                            }
+                            else {
+                                tooltip.classList.toggle("hidden")
+                            }
+                        }
```
You could extract the `showTooltip` function out of `handleFileSelect`. It looks fully generic to me.

ysbaddaden reviewed (commented):
Good job! Just a few last suggestions.

leandroradusky commented:
Waiting updates on reporting feature before merging :)
